# Tasks 



In [11]:
# Imports
import pandas as pd
import numpy as np
import sklearn as skl

### Task 1: Source the Data Set


In [12]:
# import the Iris data set from the sklearn.datasets module.
from sklearn.datasets import load_iris

# load the iris dataset
data = skl.datasets.load_iris()


#### 'Explain, in your own words, what the `load_iris()` function returns.'

The Iris data set is a famous data set used widely relating to statistics and machine learning. The data set collects a number of samples from different species of irises and their measurements between four distinct features: sepal width, sepal length, petal length and petal width. I sourced most of my information about this data set in a paper published by [**UC Irving**](https://archive.ics.uci.edu/dataset/53/iris).

The `load_iris` function from `sklearn.datasets` module returns a Bunch object, which is similar to a dictionary. It contains:

* a `.data` array with feature measurements.
* a `.target` array with numerical labels for the species.
* a `.feature_names` list with the names of the species.
* other metadata useful to understanding the data set.

This function loads the Iris dataset into memory.



### Task 2: Explore the Data Structure

#### 'Print and explain the shape of the data set, the first and last 5 rows of the data, the feature names, and the target classes.'

In [13]:
# print the shape of the data
print("Shape of dataset:", data['data'].shape)

Shape of dataset: (150, 4)


The output `(150, 4) tells us that there are 150 flower samples, which each have 4 features: sepal width, sepal length, petal length and petal width. This is the same as the UCI data set, which includes 50 samples from each of three species.

In [14]:
# create dataframe
df = pd.DataFrame(data.data, columns=data.feature_names)

# add target column
df['target'] = data.target

# display the first 5 rows of the data set
print("First 5 rows of the data set:")
print(df.head())

# display last 5 rows of the data set
print("\nLast 5 rows of the data set:")
print(df.tail())


First 5 rows of the data set:
   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0  

Last 5 rows of the data set:
     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4          

Displaying the first and last 5 rows allows me to inspect the structure and stored values in the data set. The `target` column uses integers to represent the iris species.

In [15]:
# show feature names
print("Feature names:", data.feature_names)

# show target class names
print("Target class names:", data.target_names)

# show dictionary keys
print("Keys in the data set:", data.keys())

Feature names: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
Target class names: ['setosa' 'versicolor' 'virginica']
Keys in the data set: dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


**Feature names:** are the four measurements of each iris sample

**Target class names:** the actual names of the iris species

**Keys:** The keys within the data set that help me understand it.

### Task 3: Summarize the Data

For each feature in the dataset, calculate and display:

- mean
- minimum
- maximum
- standard deviation
- median


### Task 4: Visualize Features

Plot histograms for each feature using `matplotlib`.

Add appropriate titles and axis labels.

### Task 5: Investigate Relationships

Choose any two features from the data set and create a scatter plot of them.

Color-code the three different classes of the scatter plot points.

### Task 6: Analyze Relationship

Use `numpy.polyfit` to add a regression line to the scatter plot from Task 5.

### Task 7: Analyze Class Distributions

Create box-plots of the petal lengths for each of the three classes.

### Task 8: Compute Correlations

Calculate the correlation coefficients between the features.

Display the results as a heatmap using `matplotlib`.

### Task 9: Fit a Simple Linear Regression

For your two features in Task 5, calculate the coefficient of determination $R^2$.

Re-create the plot from Task 6 and annotate it with the $R^2$ value.

### Task 10: Too Many Features

Use `seaborn` to create a `pairplot` of the data set.

Explain, in your own words, what the `pairplot` depicts.